### ISSUE TITLE
Consider introducing SPI to implement internal configuration reading in byte-buddy.

### ISSUE THREAD

**Issue by dogourd** [raphw/byte-buddy#1478]
In byte-buddy, there are switches and configuration properties to control the behavior of certain important APIs. Currently, these configurations are mainly obtained through **System.getProperty**, such as `net.bytebuddy.experimental`, `net.bytebuddy.renamed`, `net.bytebuddy.naming`, and so on. The actual situation is that there may be multiple byte-buddy dependencies in the environment, which are not shaded but work properly under different class loaders. In this case, the configuration of system properties will affect all instances of byte-buddy, causing some instances to exhibit unexpected behavior. For example, an instance that relies on the implementation by `Nexus` ends up seeing `net.bytebuddy.nexus.disabled`.

One straightforward idea is to wrap the interface for configuration reading within byte-buddy, allowing SPI to control its implementation. Of course, by default, it will use the system property-based implementation to ensure compatibility with the current behavior.

**Comment by raphw** [comment]
Well, the idea is that these switches allow disabling features that might misbehave on a given VM where it would indeed affect all users. Normally, you should not set these properties, unless you are debugging or encountering a problem. Then it would affect all uses. Where did you encounter these issues? The nexus should for example only be injected if you ever interact with the class, so it does not need to be disabled unless you actually work with it. Do you have an example where this is a problem?

**Comment by dogourd** [comment]
Currently, I haven't encountered any real-life scenarios, but I'm asking this question because I recently came across some configurations in byte-buddy. While coding, I realized that this might affect other byte-buddy instances that could exist in the environment. To put it more directly, let's say an application introduces multiple javaagents. They are all based on byte-buddy and run perfectly fine in their respective class loaders. However, if they have different configuration requirements for byte-buddy, like one javaagent wanting to skip validation of the maximum supported class file version while another javaagent wants to validate it and quickly throw an exception, system properties would enforce consistent behavior. To achieve this, we would have to delve into the internal implementation of various byte-buddy configurations and continuously modify and restore configuration properties before and after the program reads them.

The mentioned configurations are just examples, and I don't think we need to discuss the functionalities and internal implementation details of each individual configuration. The ultimate concern should be whether we have a better way to handle inconsistent byte-buddy configurations when multiple tools using byte-buddy are involved.
